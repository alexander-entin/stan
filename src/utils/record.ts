import { $global, $event, snapshot, subscribe } from '../utils/stan'

if (typeof window !== 'undefined') {
	setTimeout(() => {
		let initial
		let events = []
		window.record = () => {
			if (events.length) {
				console.log({
					initial,
					events,
				})
			}
			initial = snapshot($global)
		}
		record()
		subscribe($event, () => {
			events.push(snapshot($event))
		})
	}, 1)
}
