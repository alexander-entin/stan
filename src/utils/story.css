.diff {
	/* customization for the top-level <div> wrapper that contains the diff */
	font-family: ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
	font-size: 1em;
	white-space: pre;
}

.deletion {
	/* customization for deleted lines (<div>) */
	color: #ff0000;
}

.addition {
	/* customization for added lines (<div>) */
	color: #00dd00;
}

.unchanged {
	/* customization for unchanged lines (<div>) */
	color: gray;
}
