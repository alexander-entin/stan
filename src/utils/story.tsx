import '@unocss/reset/tailwind.css'
import 'virtual:uno.css'
import { FC, useEffect } from 'react'
import { diff } from 'deep-object-diff'
import deepRenderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'
import { deepmergeInto } from 'deepmerge-ts'

import { Event, $global, snapshot, dispatch, reset } from './stan'

export function normalize(story) {
	if (Array.isArray(story)) {
		story = { events: story }
	}
	let t = new Date('2020-02-02').valueOf()
	story.events = story.events?.map(event => {
		let id, at, type, data
		if (Array.isArray(event)) {
			[type, data] = event
		} else {
			({ id, at, type, data } = event)
		}
		id ||= `#${t}`
		at ||= t++
		data = Array.isArray(data) ? data : data === undefined ? [] : [data]
		return { id, at, type, data }
	}) || []
	return story
}

export function Story({ initial, events = [] as Event[], children }) {
	useEffect(() => {
		reset($global, initial)
		events.forEach(event => {
			const { data, ...meta } = event
			dispatch(meta)(...structuredClone(data)!)
		})
	}, [events])
	return children
}

export function tell(Component: FC, story) {
	const { initial, events } = normalize(story)
	const fs = {
		'0. initial': <Story initial={initial}><Component /></Story>
	}
	events.forEach(({ type }, idx) => {
		fs[`${idx + 1}. ${type}`] = (
			<Story initial={initial} events={events.slice(0, idx + 1)}>
				<Component />
			</Story>
		)
	})
	return fs
}

export type DeepShallow = { [method in 'deep' | 'shallow']?: Record<string, FC> }
export function test(story, components: DeepShallow, { it }) {
	let { initial = {}, events } = normalize(story)
	deepmergeInto($global, initial)
	let snap = snapshot($global)
	let { deep, shallow } = components ?? {}
	events.forEach((event, idx) => {
		const { id, at, type, data } = event
		dispatch({ id, at, type })(...(Array.isArray(data) ? data : [data]))
		const curr = snapshot($global)
		const d = diff(snap, curr)
		snap = curr
		const prefix = `${idx + 1}. ${type} >`
		it(`${prefix} $stan changes`, ({ expect }) => {
			expect(d).toMatchSnapshot()
		})
		Object.entries(shallow ?? {}).forEach(([name, Component]) => {
			const shallowRenderer = new ShallowRenderer()
			const shallow = shallowRenderer.render(<Component />)
			it(`${prefix} ${name} (shallow)`, ({ expect }) => {
				expect(shallow).toMatchSnapshot()
			})
		})
		Object.entries(deep ?? {}).forEach(([name, Component]) => {
			const deep = deepRenderer.create(<Component />).toJSON()
			it(`${prefix} ${name} (deep)`, ({ expect }) => {
				expect(deep).toMatchSnapshot()
			})
		})
	})
}
