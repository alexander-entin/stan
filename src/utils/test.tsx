import { describe, it } from 'vitest'
import { diff } from 'deep-object-diff'
import deepRenderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'

import { dispatch, $global, snapshot } from './stan'

export function test(cases, defaultComponents) {
	Object.entries(cases).forEach(([name, { events, components }]) => {
		components = components || defaultComponents
		describe(name, () => {
			let t = new Date('2020-02-02').valueOf()
			let snap = snapshot($global)
			events.forEach((event, idx) => {
				const { id = `#${t}`, at = t, type, data } = event
				dispatch({ id, at, type })(...(Array.isArray(data) ? data : [data]))
				const curr = snapshot($global)
				const d = diff(snap, curr)
				delete d.event
				snap = curr
				it(`${idx}. ${type} > $stan changes`, ({ expect }) => {
					expect(d).toMatchSnapshot()
				})
				const cs = typeof components === 'function' ? components(event, idx) : components
				if (cs) {
					Object.entries(cs).forEach(([name, Component]) => {
						const component = <Component />
						const shallowRenderer = new ShallowRenderer()
						const shallow = shallowRenderer.render(component)
						it(`${idx}. ${type} > ${name} (shallow)`, ({ expect }) => {
							expect(shallow).toMatchSnapshot()
						})
						const deep = deepRenderer.create(component).toJSON()
						it(`${idx}. ${type} > ${name} (deep)`, ({ expect }) => {
							expect(deep).toMatchSnapshot()
						})
					})
				}
				t++
			})
		})
	})
}