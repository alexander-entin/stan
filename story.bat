start "story" /max pnpm storybook
