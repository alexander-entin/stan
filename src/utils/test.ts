import { it } from 'vitest'

import { dispatch, $global, snapshot } from '../utils/stan'

export function test($stan, cases) {
	Object.entries(cases).forEach(([name, spec]) => {
		it(name, ({ expect }) => {
			spec.events.forEach(({ data, ...meta }) =>
				dispatch(meta)(...(Array.isArray(data) ? data : [data]))
			)
			const snap = snapshot($global)
			expect(snap).toMatchSnapshot()
		})
	})
}