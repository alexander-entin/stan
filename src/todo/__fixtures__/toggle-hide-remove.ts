import { tell, test } from '/src/utils/story'
import Page from '../+Page'

const story = [
	["todo.onPull", { "pqRxsRzfIvfPhyL1Ol7cb": { "id": "pqRxsRzfIvfPhyL1Ol7cb", "text": "read", "done": false }, "XONBkwGngyqr60paySNfs": { "id": "XONBkwGngyqr60paySNfs", "text": "buy milk", "done": false }, "yFtPqsesTs2Q6IcJS3-cG": { "id": "yFtPqsesTs2Q6IcJS3-cG", "text": "buy meat" } }],
	["todo.onToggle", "XONBkwGngyqr60paySNfs"],
	["todo.onFilter", "!done"],
	["todo.onDeleteDone"],
	["todo.onFilter", ""],
	["todo.onDelete", "yFtPqsesTs2Q6IcJS3-cG"],
]

export default tell(Page, story)

if (import.meta.vitest) {
	test(story, { deep: { Page } }, import.meta.vitest)
}
