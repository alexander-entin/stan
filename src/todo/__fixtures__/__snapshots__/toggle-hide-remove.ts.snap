// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`1. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": "read "Out of the Tar Pit" by Ben Moseley and Peter Marks",
  },
}
`;

exports[`1. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value="read "Out of the Tar Pit" by Ben Moseley and Peter Marks"
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`1. todo-local.onPull > $stan changes 1`] = `
{
  "todo": {
    "filtered": {
      "0": {
        "done": false,
        "id": "pqRxsRzfIvfPhyL1Ol7cb",
        "text": "read",
      },
      "1": {
        "done": true,
        "id": "XONBkwGngyqr60paySNfs",
        "text": "buy milk",
      },
      "2": {
        "id": "yFtPqsesTs2Q6IcJS3-cG",
        "text": "buy meat",
      },
    },
    "map": {
      "XONBkwGngyqr60paySNfs": {
        "done": false,
        "id": "XONBkwGngyqr60paySNfs",
        "text": "buy milk",
      },
      "pqRxsRzfIvfPhyL1Ol7cb": {
        "done": false,
        "id": "pqRxsRzfIvfPhyL1Ol7cb",
        "text": "read",
      },
      "yFtPqsesTs2Q6IcJS3-cG": {
        "id": "yFtPqsesTs2Q6IcJS3-cG",
        "text": "buy meat",
      },
    },
  },
}
`;

exports[`1. todo-local.onPull > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`1. todo-ui.onPull > $stan changes 1`] = `{}`;

exports[`1. todo-ui.onPull > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`2. todo.onAdd > $stan changes 1`] = `
{
  "todo": {
    "draft": "",
    "filtered": {
      "0": {
        "id": "#1580601600001",
        "text": "read "Out of the Tar Pit" by Ben Moseley and Peter Marks",
      },
    },
    "map": {
      "#1580601600001": {
        "id": "#1580601600001",
        "text": "read "Out of the Tar Pit" by Ben Moseley and Peter Marks",
      },
    },
  },
}
`;

exports[`2. todo.onAdd > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      read "Out of the Tar Pit" by Ben Moseley and Peter Marks
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`2. todo.onToggle > $stan changes 1`] = `
{
  "todo": {
    "map": {
      "XONBkwGngyqr60paySNfs": {
        "done": true,
      },
    },
  },
}
`;

exports[`2. todo.onToggle > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      read
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
    <li
      className="line-through"
    >
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
      />
      buy milk
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy meat
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`2. todo-local.onPull > $stan changes 1`] = `
{
  "todo": {
    "filtered": {
      "0": {
        "done": false,
        "id": "pqRxsRzfIvfPhyL1Ol7cb",
        "text": "read",
      },
      "1": {
        "done": true,
        "id": "XONBkwGngyqr60paySNfs",
        "text": "buy milk",
      },
      "2": {
        "id": "yFtPqsesTs2Q6IcJS3-cG",
        "text": "buy meat",
      },
    },
    "map": {
      "XONBkwGngyqr60paySNfs": {
        "done": false,
        "id": "XONBkwGngyqr60paySNfs",
        "text": "buy milk",
      },
      "pqRxsRzfIvfPhyL1Ol7cb": {
        "done": false,
        "id": "pqRxsRzfIvfPhyL1Ol7cb",
        "text": "read",
      },
      "yFtPqsesTs2Q6IcJS3-cG": {
        "id": "yFtPqsesTs2Q6IcJS3-cG",
        "text": "buy meat",
      },
    },
  },
}
`;

exports[`2. todo-local.onPull > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      read
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy milk
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy meat
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`3. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": "buy milk",
    "filtered": {
      "0": undefined,
    },
  },
}
`;

exports[`3. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value="buy milk"
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul />,
]
`;

exports[`3. todo.onFilter > $stan changes 1`] = `
{
  "todo": {
    "filter": "!done",
    "filtered": {
      "1": {
        "done": undefined,
        "id": "yFtPqsesTs2Q6IcJS3-cG",
        "text": "buy meat",
      },
      "2": undefined,
    },
  },
}
`;

exports[`3. todo.onFilter > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`3. todo.onToggle > $stan changes 1`] = `
{
  "todo": {
    "map": {
      "XONBkwGngyqr60paySNfs": {
        "done": true,
      },
    },
  },
}
`;

exports[`3. todo.onToggle > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`4. todo.onAdd > $stan changes 1`] = `
{
  "todo": {
    "draft": "",
    "filtered": {
      "0": {
        "id": "#1580601600001",
        "text": "read "Out of the Tar Pit" by Ben Moseley and Peter Marks",
      },
      "1": {
        "id": "#1580601600003",
        "text": "buy milk",
      },
    },
    "map": {
      "#1580601600003": {
        "id": "#1580601600003",
        "text": "buy milk",
      },
    },
  },
}
`;

exports[`4. todo.onAdd > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`4. todo.onDeleteDone > $stan changes 1`] = `
{
  "todo": {
    "map": {
      "XONBkwGngyqr60paySNfs": undefined,
    },
  },
}
`;

exports[`4. todo.onDeleteDone > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={true}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      read
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy meat
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`4. todo.onFilter > $stan changes 1`] = `
{
  "todo": {
    "filter": "!done",
    "filtered": {
      "1": {
        "done": undefined,
        "id": "yFtPqsesTs2Q6IcJS3-cG",
        "text": "buy meat",
      },
      "2": undefined,
    },
  },
}
`;

exports[`4. todo.onFilter > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={true}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      read
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy meat
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`5. todo.onAdd > $stan changes 1`] = `{}`;

exports[`5. todo.onAdd > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      read "Out of the Tar Pit" by Ben Moseley and Peter Marks
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy milk
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`5. todo.onDeleteDone > $stan changes 1`] = `
{
  "todo": {
    "map": {
      "XONBkwGngyqr60paySNfs": undefined,
    },
  },
}
`;

exports[`5. todo.onDeleteDone > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`5. todo.onFilter > $stan changes 1`] = `
{
  "todo": {
    "filter": "",
  },
}
`;

exports[`5. todo.onFilter > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`6. todo.onDelete > $stan changes 1`] = `
{
  "todo": {
    "filtered": {
      "1": undefined,
    },
    "map": {
      "yFtPqsesTs2Q6IcJS3-cG": undefined,
    },
  },
}
`;

exports[`6. todo.onDelete > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      read
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`6. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": " ",
    "filtered": {
      "0": undefined,
      "1": undefined,
    },
  },
}
`;

exports[`6. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=" "
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul />,
]
`;

exports[`6. todo.onFilter > $stan changes 1`] = `
{
  "todo": {
    "filter": "",
  },
}
`;

exports[`6. todo.onFilter > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      read
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy meat
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`7. todo.onAdd > $stan changes 1`] = `{}`;

exports[`7. todo.onAdd > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=" "
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`7. todo.onDelete > $stan changes 1`] = `
{
  "todo": {
    "filtered": {
      "1": undefined,
    },
    "map": {
      "yFtPqsesTs2Q6IcJS3-cG": undefined,
    },
  },
}
`;

exports[`7. todo.onDelete > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`8. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": "b",
    "filtered": {
      "0": {
        "id": "#1580601600003",
        "text": "buy milk",
      },
    },
  },
}
`;

exports[`8. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value="b"
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy milk
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`9. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": "bu",
  },
}
`;

exports[`9. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value="bu"
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy milk
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`10. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": "buy",
  },
}
`;

exports[`10. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value="buy"
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`11. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": "buy ",
  },
}
`;

exports[`11. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value="buy "
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy milk
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`12. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": "buy m",
  },
}
`;

exports[`12. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value="buy m"
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul>
    <li
      className=""
    >
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      buy milk
       
      <button
        onClick={[Function]}
      >
        ☠
      </button>
    </li>
  </ul>,
]
`;

exports[`13. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": "buy me",
    "filtered": {
      "0": undefined,
    },
  },
}
`;

exports[`13. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value="buy me"
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;

exports[`14. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": "buy mea",
  },
}
`;

exports[`14. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value="buy mea"
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul />,
]
`;

exports[`15. todo.onDraft > $stan changes 1`] = `
{
  "todo": {
    "draft": "buy meat",
  },
}
`;

exports[`15. todo.onDraft > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value="buy meat"
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  "Hide completed",
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    Remove completed
  </button>,
  <ul />,
]
`;

exports[`16. todo.onAdd > $stan changes 1`] = `
{
  "todo": {
    "draft": "",
    "filtered": {
      "0": {
        "id": "#1580601600001",
        "text": "read "Out of the Tar Pit" by Ben Moseley and Peter Marks",
      },
      "1": {
        "id": "#1580601600003",
        "text": "buy milk",
      },
      "2": {
        "id": "#1580601600015",
        "text": "buy meat",
      },
    },
    "map": {
      "#1580601600015": {
        "id": "#1580601600015",
        "text": "buy meat",
      },
    },
  },
}
`;

exports[`16. todo.onAdd > Page (deep) 1`] = `
[
  <h1>
    To-do List
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <input
      onInput={[Function]}
      placeholder="New Todo"
      type="text"
      value=""
    />
    <button
      className="btn"
      type="submit"
    >
      Create
    </button>
  </form>,
]
`;
